### PR TITLE
aufs: apply dirperm1 by default if supported

### LINKED ISSUE #783: unexpected file permission error in container
I've narrowed down a problem in an originally more involved setup. Consider the following Dockerfile:

    # Dockerfile
    FROM      ubuntu:12.10

    RUN apt-get install -y puppetmaster sudo

    RUN rm -rf /etc/puppet
    ADD puppet-config /etc/puppet
    RUN chown -R puppet.puppet /etc/puppet
    RUN chmod 755 /etc/puppet

When run with the following:

    # make a dummy directory
    mkdir puppet-config
    echo "hi" >puppet-config/hello.txt

    docker build -t dockbug .

    echo "note the directory is owned by puppet with full read/write/execute privs"
    docker run dockbug ls -al /etc/puppet

    echo "but we get a permission error here"
    docker run dockbug sudo -u puppet ls -al /etc/puppet

I see an unexpected permission error in the final command. This is with Docker 0.3.4 from the PPA on Ubuntu 13.04 with kernel 3.8.0-19-generic. Interestingly, if I remove the like "RUN rm -rf /etc/puppet" from the Dockerfile, I no longer see the permission error.

### PR DESCRIPTION
**Pull request by dqminh**
Automatically detect support for aufs `dirperm1` option and apply it.
`dirperm1` tells aufs to check the permission bits of the directory on the
topmost branch and ignore the permission bits on all lower branches.
It can be used to fix aufs' permission bug (i.e., upper layer having
broader mask than the lower layer).

`dirperm1` man page is at: http://aufs.sourceforge.net/aufs3/man.html

Fixes #783

### CODE REVIEW

**Comment by dqminh** [comment]
cc @tiborvass 

Also it seems like enabling `dirperm1` makes remounting layer significantly slower (1+ms in my VM) ( happens when the user has a lot of layers that total length of mount options exceeds PAGE_SIZE ). `TestMountMoreThan42Layers` is much more slower with this patch.

**Comment by tiborvass** [comment]
+1 for this, I like it.

Ping @vbatts @unclejack feel free to step in to say if you don't like the way this is implemented.

File: daemon/graphdriver/aufs/aufs.go [hunk 1]
```diff
@@ -460,3 +472,36 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
...
+// useDirperm checks dirperm1 mount option can be used with the current
+// version of aufs.
+func useDirperm() bool {
+	enableDirpermLock.Do(func() {
+		var (
+			base, union string
+			err         error
+		)
+
+		if base, err = ioutil.TempDir("", "docker-aufs-base"); err != nil {
```
Comment (tiborvass) [inline daemon/graphdriver/aufs/aufs.go]: should be 
```Go
base, err := ioutil.TempDir("", "docker-aufs-base")
if err != nil {
```

Same for `union, err :=`

File: daemon/graphdriver/aufs/aufs.go [hunk 2]
```diff
@@ -460,3 +472,36 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
...
+		defer os.RemoveAll(base)
+
+		if union, err = ioutil.TempDir("", "docker-aufs-union"); err != nil {
+			log.Errorf("error checking dirperm1: %v", err)
+			return
+		}
+		defer os.RemoveAll(union)
+
+		opts := fmt.Sprintf("br:%s,dirperm1,xino=/dev/shm/aufs.xino", base)
+		if err = mount("none", union, "aufs", 0, opts); err != nil {
```
Comment (tiborvass) [inline daemon/graphdriver/aufs/aufs.go]: `err :=`

**Comment by tiborvass** [comment]
@jfrazelle this would be a great fix that a lot of people would welcome. Also it would close an issue with a 3-digit number :P

@dqminh would you mind removing the part that mentions the issue in some KNOWN issues section somewhere?

**Comment by dqminh** [comment]
> would you mind removing the part that mentions the issue in some KNOWN issues section somewhere?

@tiborvass this won't fix it for system that doesnt have `dirperm1` patch though, so I think we still need to keep the note. I can add some note about `dirperm1` being set as the default option if it's available.

**Comment by tiborvass** [comment]
@dqminh fair enough! Thanks :)

**Comment by crosbymichael** [comment]
LGTM

**Comment by dqminh** [comment]
@tiborvass i added dirperm1 support status to `docker info` output too so there's an easy way to tell user whether their system is still affected by the permission bug.

**Comment by dqminh** [comment]
@tiborvass is this merge-able for 1.6, or is it too late now ?

**Comment by jfrazelle** [comment]
I added the milestone @crosbymichael is that ok with you

**Comment by vbatts** [comment]
LGTM

File: daemon/graphdriver/aufs/aufs.go [hunk 3]
```diff
@@ -460,3 +473,33 @@ func (a *Driver) aufsMount(ro []string, rw, target, mountLabel string) (err erro
...
+
+		union, err := ioutil.TempDir("", "docker-aufs-union")
+		if err != nil {
+			log.Errorf("error checking dirperm1: %v", err)
+			return
+		}
+		defer os.RemoveAll(union)
+
+		opts := fmt.Sprintf("br:%s,dirperm1,xino=/dev/shm/aufs.xino", base)
+		if err := mount("none", union, "aufs", 0, opts); err != nil {
```
Comment (Blaisorblade) [inline daemon/graphdriver/aufs/aufs.go]: While investigating some other (transient) docker problem, I saw this error (once) in my kernel log — is it a harmless error produced here while testing the presence of `dirperm1`? Google wasn't very helpful.
```
[   16.825747] aufs au_opts_parse:1155:docker[1679]: unknown option dirperm1
```

Comment (Blaisorblade) [inline daemon/graphdriver/aufs/aufs.go]: OK, I'm guessing yes:
```
# docker info | grep Dirperm1
 Dirperm1 Supported: false
```

**Pull request merged**